File: backend/src/controllers/opportunitiesController.js
const { validationResult } = require('express-validator');
const Opportunity = require('../models/Opportunity');

const getOpportunities = async (req, res, next) => {
  try {
    const { lead_id, stage, agent_id, page = 1, limit = 20 } = req.query;

    const filters = {};
    if (lead_id) filters.lead_id = parseInt(lead_id);
    if (stage) filters.stage = stage;

    const { role_name, user_id } = req.user;
    if (role_name === 'Sales Agent') {
      filters.agent_id = user_id;
    } else if (agent_id) {
      filters.agent_id = parseInt(agent_id);
    }

    const pagination = { page: parseInt(page), limit: Math.min(parseInt(limit), 100) };
    const { data, total } = await Opportunity.findAll(filters, pagination);

    return res.status(200).json({
      success: true,
      data,
      pagination: {
        page: pagination.page,
        limit: pagination.limit,
        total,
        totalPages: Math.ceil(total / pagination.limit)
      }
    });
  } catch (err) {
    next(err);
  }
};

const createOpportunity = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const { lead_id, stage, premium_amount, probability, notes } = req.body;

    const opportunity = await Opportunity.create({ lead_id, stage, premium_amount, probability, notes });

    res.locals.createdId = opportunity.opportunity_id;
    return res.status(201).json({ success: true, data: opportunity });
  } catch (err) {
    next(err);
  }
};

const getPipeline = async (req, res, next) => {
  try {
    const { role_name, user_id } = req.user;
    const filters = {};
    if (role_name === 'Sales Agent') {
      filters.agent_id = user_id;
    } else if (req.query.agent_id) {
      filters.agent_id = parseInt(req.query.agent_id);
    }

    const data = await Opportunity.getPipeline(filters);
    return res.status(200).json({ success: true, data });
  } catch (err) {
    next(err);
  }
};

const updateOpportunity = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const id = parseInt(req.params.id);
    const existing = await Opportunity.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'Opportunity not found' });
    }

    const updated = await Opportunity.update(id, req.body);
    return res.status(200).json({ success: true, data: updated });
  } catch (err) {
    next(err);
  }
};

const deleteOpportunity = async (req, res, next) => {
  try {
    const id = parseInt(req.params.id);
    const existing = await Opportunity.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'Opportunity not found' });
    }

    await Opportunity.remove(id);
    return res.status(200).json({ success: true, message: 'Opportunity deleted successfully' });
  } catch (err) {
    next(err);
  }
};

module.exports = { getOpportunities, createOpportunity, getPipeline, updateOpportunity, deleteOpportunity };

File: dotnet-api/Controllers/OpportunitiesController.cs
using ActivityTrackerAPI.Helpers;
using ActivityTrackerAPI.Models.DTOs;
using ActivityTrackerAPI.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ActivityTrackerAPI.Controllers;

/// <summary>Sales opportunity management endpoints</summary>
[ApiController]
[Route("api/opportunities")]
//[Authorize]
[Produces("application/json")]
public class OpportunitiesController : ControllerBase
{
    private readonly IOpportunityService _opportunityService;

    public OpportunitiesController(IOpportunityService opportunityService)
    {
        _opportunityService = opportunityService;
    }

    /// <summary>List opportunities with optional filters and pagination</summary>
    [HttpGet]
    [ProducesResponseType(200)]
    public async Task<IActionResult> GetOpportunities(
        [FromQuery] uint? lead_id,
        [FromQuery] string? stage,
        [FromQuery] uint? agent_id,
        [FromQuery] int page = 1,
        [FromQuery] int limit = 20)
    {
        var roleName = User.GetRoleName();
        var userId = User.GetUserId();

        uint? filterAgentId = agent_id;
        if (roleName == "Sales Agent")
            filterAgentId = userId;

        limit = Math.Min(limit, 100);
        var (data, total) = await _opportunityService.GetAllAsync(lead_id, stage, filterAgentId, page, limit);

        return Ok(new
        {
            success = true,
            data,
            pagination = new { page, limit, total, totalPages = (int)Math.Ceiling((double)total / limit) }
        });
    }

    /// <summary>Create a new opportunity</summary>
    [HttpPost]
    [ProducesResponseType(201)]
    [ProducesResponseType(400)]
    public async Task<IActionResult> CreateOpportunity([FromBody] CreateOpportunityRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var opp = await _opportunityService.CreateAsync(
            request.LeadId, request.Stage, request.PremiumAmount, request.Probability, request.Notes);

        return StatusCode(201, new { success = true, data = opp });
    }

    /// <summary>Get aggregated pipeline by stage</summary>
    [HttpGet("pipeline")]
    [ProducesResponseType(200)]
    public async Task<IActionResult> GetPipeline([FromQuery] uint? agent_id)
    {
        var roleName = User.GetRoleName();
        var userId = User.GetUserId();

        uint? filterAgentId = agent_id;
        if (roleName == "Sales Agent")
            filterAgentId = userId;

        var data = await _opportunityService.GetPipelineAsync(filterAgentId);
        return Ok(new { success = true, data });
    }

    /// <summary>Update an opportunity</summary>
    [HttpPut("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> UpdateOpportunity(uint id, [FromBody] UpdateOpportunityRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var existing = await _opportunityService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "Opportunity not found" });

        var updated = await _opportunityService.UpdateAsync(id,
            request.Stage, request.PremiumAmount, request.Probability, request.Notes);

        return Ok(new { success = true, data = updated });
    }

    /// <summary>Delete an opportunity</summary>
    [HttpDelete("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> DeleteOpportunity(uint id)
    {
        var existing = await _opportunityService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "Opportunity not found" });

        await _opportunityService.DeleteAsync(id);
        return Ok(new { success = true, message = "Opportunity deleted successfully" });
    }
}
